src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1849,6 +1890,49 @@ private void RemoveStream(Http2Stream http2Stream)
             _concurrentStreams.AdjustCredit(1);
         }
 
+        private void RefreshPingTimestamp()
+        {
+            _nextPingRequestTimestamp = Environment.TickCount64 + (long)_keepAlivePingDelay.TotalMilliseconds;
+        }
+
+        private void ProcessPingAck(long payload)
+        {
+            if (_keepAliveState != KeepAliveState.PingSent)
+                ThrowProtocolError();
+            if (Interlocked.Read(ref _keepAlivePingPayload) != payload)
+                ThrowProtocolError();
+            _keepAliveState = KeepAliveState.None;
+            RefreshPingTimestamp();
+        }
+
+        private void VerifyKeepAlive()
+        {
+            if (_keepAlivePingPolicy == HttpKeepAlivePingPolicy.WithActiveRequests && _httpStreams.Count == 0)
+                return;
+
+            var now = Environment.TickCount64;
+            switch (_keepAliveState)
+            {
+                case KeepAliveState.None:
+                    // Check whether keep alive delay has passed since last frame received
+                    if (_keepAlivePingDelay != Timeout.InfiniteTimeSpan && now > _nextPingRequestTimestamp)
+                    {
+                        // Set the status directly to ping sent and set the timestamp
+                        _keepAliveState = KeepAliveState.PingSent;
+                        _keepAlivePingTimeoutTimestamp = now + (long)_keepAlivePingTimeout.TotalMilliseconds;
+
+                        long pingPayload = Interlocked.Increment(ref _keepAlivePingPayload);
```

Comment:
Out of interest, is there a reason you want to send each ping with a unique payload and verify it?

Reply:
The reason is just to identify the specific ping ack to ensure server response correctly.